### ISSUE TITLE
Allow for different handling of OOV words

### ISSUE THREAD

**Issue by matt-gardner** [allenai/allennlp#239]
We currently give all OOV tokens the same embedding at both training and test time.  It'd be nice to be able to have some different options here:

- At test time, see if the OOV token is in glove, and use that embedding instead
- At both training and test, use random vectors for each unique OOV token, as suggested [here](https://www.semanticscholar.org/paper/A-Comparative-Study-of-Word-Embeddings-for-Reading-Dhingra-Liu/3ec37205c9201fc891ab51da200e361fdc34bfb3)

There are probably some other options I'm forgetting right now.  These would be pretty tricky to implement in our current data pipeline, though.

**Comment by matt-gardner** [comment]
@OyvindTafjord made some nice progress on getting some of this to work:

https://github.com/allenai/allennlp/compare/master...OyvindTafjord:expandVocab

Here are Oyvind's notes on this diff:

  * If you use --restore in the train command, it reads the vocabulary from the stored model rather than from the data files (essential if you're continuing training with a different dataset than the original model).

  * If you use --expand_vocabulary (together with --restore) in the train command, it will expand the embedding matrix with any new words found in the current dataset. This is done using a very hacky hook to the trainer, and assuming the BiDAF structure of the text field embedder, so this will be broken in general. It saves the new vocabulary in a vocabulary_new directory, so to get a proper model.tar.gz at the end, you have to manually move it to vocabulary and rerun the archive generation.mic

  * You can also use --expand_vocabulary when doing an evaluation to the same effect.

  * I did not do anything special about the training_state recovery, it might make sense to give some more control of that (e.g.,. not require a training state, or allow to start from original state rather than the current state).

  * I also added --output_file to the evaluate command, but this ended up being sufficiently ugly that there should probably be a more general way to get custom output for a run, rather than having to monkey around inside the forward method.

**Comment by matt-gardner** [comment]
@DeNeutoy, @joelgrus, I think it probably makes sense to have a `load_from` argument to trainer that's separate from the `serialization_dir`, instead of how we're currently looking at the serialization dir and continuing if it's not empty.  Or maybe we should be able to just specify a model archive location for the `model` parameter in the config file?  We don't have a great story around fine-tuning models.

If you're loading a saved model, you should _always_ expand the vocabulary with the new dataset in this way (or in a way that's configurable).

Ok, so I think there are three separate issues here:

1. We should have a better way to do model fine-tuning.
2. We should always expand the vocabulary on evaluate (or when fine-tuning a model).
3. We should make the behavior of OOV words configurable (random vector vs. UNK token vs. ...).

Let's keep this issue tracking (3), and I'll open new issues for (1) and (2), referencing this one.

**Comment by ngarneau** [comment]
Hey there,
I have a question about OOV words and I think it is related to this issue. I just clone your repo and gave a ran to the crf_tagger model. I saw that it is using the test set to build its vocabulary. Isn't it like peeking into the test dataset? In that case, the only OOV words that you will have will be words that are not into the Glove embeddings.
Please correct me if I'm missing something here.
Thanks

**Comment by matt-gardner** [comment]
Yes, we have an option to use the test set to compute the vocabulary.  This is so that you can use glove embeddings for words that show up only in the test set.  This isn't cheating, because words that are added this way are never seen at training time and thus never get trained - you're not getting any signal from the test set.

**Comment by ngarneau** [comment]
Great, I didn't see it that way but it is definitely correct. In the case you'd be interested, I am looking to incorporate a way to generate meaningful embeddings for OOV words, based on this paper: http://www.aclweb.org/anthology/D17-1010

The basic idea is to train a character lstm over the Glove embeddings so that you can get meaningful word embeddings for words you haven't seen, based on their characters.

I would like to add this into the crf_tagger model.

**Comment by matt-gardner** [comment]
You can use a character-level CNN or LSTM right now, without writing any additional code.  If you want train it to match glove embeddings, use pre-trained weights for it, or only use it for OOV words instead of all words, those are all a little more tricky, but should still be possible.
